translator/cmdutil/userutil_linux.go

```diff
@@ -6,29 +6,108 @@
 package cmdutil
 
 import (
+	"bufio"
+	"bytes"
 	"fmt"
+	"golang.org/x/sys/unix"
+	"io"
 	"log"
 	"os"
-	"syscall"
+	"os/user"
+	"strconv"
+	"strings"
+)
 
-	"golang.org/x/sys/unix"
+type ExecUser struct {
+	Uid  int
+	Gid  int
+	Home string
+	Gids []int
+}
 
-	"github.com/opencontainers/runc/libcontainer/system"
-	"github.com/opencontainers/runc/libcontainer/user"
-)
+func getGroupIds(u *user.User) ([]int, error) {
+	file, err := os.Open("/etc/group")
+	if err != nil {
+		log.Printf("E! Failed to open group file: %v", err)
+		return nil, err
+	}
+	defer file.Close()
+	reader := bufio.NewReader(file)
+	groupIds := []int{}
+	for {
+		var wholeLine []byte
+		for {
+			line, isPrefix, err := reader.ReadLine()
+
+			if err != nil {
+				// EOF reached
+				if err == io.EOF {
+					return groupIds, nil
+				}
+				return nil, err
+			}
+
+			// Whole line was able to fit in single buffer
+			if !isPrefix && len(wholeLine) == 0 {
+				wholeLine = line
+				break
+			}
+			wholeLine = append(wholeLine, line...)
+			// Last fragment of line read
+			if !isPrefix {
+				break
+			}
+		}
+
+		wholeLine = bytes.TrimSpace(wholeLine)
+		// Not empty, not a comment, and has enough parts
+		if len(wholeLine) == 0 || wholeLine[0] == '#' || bytes.Count(wholeLine, []byte{':'}) < 3 {
+			continue
+		}
+		parts := strings.SplitN(string(wholeLine), ":", 4)
+		users := strings.TrimSpace(parts[3])
+		if len(users) == 0 || !strings.Contains(users, u.Username) {
+			continue
+		}
+		groupId, err := strconv.Atoi(parts[2])
+		if err != nil {
+			return nil, err
+		}
+		groupIds = append(groupIds, groupId)
+	}
+}
+
+func toExecUser(u *user.User) (*ExecUser, error) {
+	uid, err := strconv.Atoi(u.Uid)
+	if err != nil {
+		log.Printf("E! Failed to convert uid to int: %v", err)
+		return nil, err
+	}
+	gid, err := strconv.Atoi(u.Gid)
+	if err != nil {
+		log.Printf("E! Failed to convert gid to int: %v", err)
+		return nil, err
+	}
+	gids, err := getGroupIds(u)
```

Comment:
Was there a reason why you couldn't use `u.GroupIds()` in this case?

Reply:
The `GroupIds` function introduces a dependency on cgo, which can be enabled with `CGO_ENABLED=1`, but adds a lot of unnecessary build complexity for a single function.

https://pkg.go.dev/cmd/cgo
>When cross-compiling, you must specify a C cross-compiler for cgo to use. You can do this by setting the generic CC_FOR_TARGET or the more specific CC_FOR_${GOOS}_${GOARCH} (for example, CC_FOR_linux_arm) environment variable when building the toolchain using make.bash, or you can set the CC environment variable any time you run the go tool.